src/modules/sdexec/sdexec.c

```diff
@@ -0,0 +1,850 @@
+    while (request) {
+        if (flux_disconnect_match (msg, request)) {
+            struct sdproc *proc = flux_msg_aux_get (request, "sdproc");
+            if (proc) {
+                flux_future_t *f;
+                f = sdexec_kill_unit (h,
+                                      ctx->rank,
+                                      sdexec_unit_name (proc->unit),
+                                      "main",
+                                       signum);
+                flux_future_destroy (f);
+            }
+        }
+        request = flux_msglist_next (ctx->requests);
+    }
+}
+
+static struct flux_msg_handler_spec htab[] = {
+    { FLUX_MSGTYPE_REQUEST,
+      "disconnect",
+      disconnect_cb,
+      0
+    },
+    { FLUX_MSGTYPE_REQUEST,
+      "exec",
+      exec_cb,
+      0
+    },
+    { FLUX_MSGTYPE_REQUEST,
+      "write",
+      write_cb,
+      0
+    },
+    { FLUX_MSGTYPE_REQUEST,
+      "kill",
+      kill_cb,
+      0
+    },
+    FLUX_MSGHANDLER_TABLE_END
+};
+
+/* N.B. systemd.enable is tested in rc1 and ignored here.
+ */
+static int parse_config (flux_t *h, const char **unit_prefix)
+{
+    flux_error_t error;
+    int enable = false;
+    const char *s = "flux-";
+
+    if (flux_conf_unpack (flux_get_conf (h),
+                          &error,
+                          "{s?{s?b s?s}}",
+                          "systemd",
+                            "enable", &enable,
+                            "unit-prefix", &s) < 0) {
+        flux_log (h, LOG_ERR, "error parsing [systemd] config: %s", error.text);
+        return -1;
+    }
+    *unit_prefix = s;
+    return 0;
+}
+
+static void sdexec_ctx_destroy (struct sdexec_ctx *ctx)
+{
+    if (ctx) {
+        int saved_errno = errno;
+        flux_msg_handler_delvec (ctx->handlers);
+        if (ctx->requests) {
+            const flux_msg_t *msg;
+            msg = flux_msglist_first (ctx->requests);
+            while (msg) {
+                const char *errstr = "sdexec module is unloading";
+                if (flux_respond_error (ctx->h, msg, ENOSYS, errstr) < 0)
+                    flux_log_error (ctx->h, "error responding to exec request");
+                msg = flux_msglist_next (ctx->requests);
+            }
+            flux_msglist_destroy (ctx->requests);
+        }
+        free (ctx->unit_prefix);
+        flux_msglist_destroy (ctx->kills);
+        free (ctx->local_uri);
+        free (ctx);
+        errno = saved_errno;
+    }
+}
+
+static struct sdexec_ctx *sdexec_ctx_create (flux_t *h)
+{
+    struct sdexec_ctx *ctx;
+    const char *s;
+
+    if (!(ctx = calloc (1, sizeof (*ctx))))
+        return NULL;
+    ctx->h = h;
+    if (flux_get_rank (h, &ctx->rank) < 0)
+        goto error;
+    /* Add the broker rank to the unit prefix to avoid name collisions
+     * in test instances.
+     */
+    if (parse_config (h, &s) < 0
+        || asprintf (&ctx->unit_prefix, "%s%d-", s, (int)ctx->rank) < 0)
```

Comment:
noticed you put "flux-" as default, but would "flux" be more reasonable and you put a "-" here?  i.e. `%s-%d`.  Doubt average user would think to put "-" in the config.

Reply:
Good point. Still pondering how best to manage the clean up of lingering process so holding off on changing anything in that area of code for this iteration though.

Reply:
In the latest push I eliminated the configured unit prefix.  If an exec request does not have the SDEXEC_NAME option set, then a UUID is used.  Otherwise, it is used verbatim.  Then in `job-exec`, SDEXEC_NAME is set to either `shell-N-jobid` or `imp-kill-N-jobid` (where N is the broker rank).